### PR TITLE
add search and ordering in courses and users general views

### PR DESCRIPTION
**Pull request by melvinsoft**
From @grozdanowski request:

What’s needed:

* filtering: 
** `/figures/api/users/general/` - I want to filter by either username, fullname, email (if any of these contains string)
** `/figures/api/courses/general/` - I want to filter by either course_name, course_id, email (if any of these contains string)

* sorting - I want to sort the returned list:
** `/figures/api/users/general/`-Sort by (alphabetically) username, email, fullname; is_active
** `/figures/api/courses/general/` - Sort by (alphabetically) course name; (boolean) self-paced; (numerical) num_learners_completed, enrollment_count

Sorting needs to be able to be A->Z / Z->A; 1->9 / 9->1

The only functionality I excluded is to sort by num_learners_completed, it's really hard to turn that relationship value into an annotated value, if it adds a good value, we can invest the time in the near future.

### CODE REVIEW

File: figures/filters.py [hunk 1]
```diff
@@ -116,6 +116,7 @@ class UserFilterSet(django_filters.FilterSet):
     is_superuser = django_filters.BooleanFilter(name='is_superuser')
     username = django_filters.CharFilter(lookup_expr='icontains')
     email = django_filters.CharFilter(lookup_expr='icontains')
+    profile__name = django_filters.CharFilter(lookup_expr='icontains')
```
Comment (OmarIthawi) [inline figures/filters.py]:
```suggestion
    name = django_filters.CharFilter(lookup_expr='icontains', name='profile__name')
```

File: figures/filters.py [hunk 2]
```diff
@@ -124,7 +125,7 @@ class UserFilterSet(django_filters.FilterSet):
 
     class Meta:
         model = get_user_model()
-        fields = ['username', 'email', 'country', 'is_active', 'is_staff',
+        fields = ['username', 'email', 'profile__name', 'country', 'is_active', 'is_staff',
```
Comment (OmarIthawi) [inline figures/filters.py]:
```suggestion
        fields = ['username', 'email', 'name', 'country', 'is_active', 'is_staff',
```

File: figures/views.py [hunk 3]
```diff
@@ -292,6 +296,10 @@ class GeneralCourseDataViewSet(CommonAuthMixin, viewsets.ReadOnlyModelViewSet):
     # have to change the front end until Figures "Level 2"
     pagination_class = FiguresKiloPagination
     serializer_class = GeneralCourseDataSerializer
+    filter_backends = (SearchFilter, DjangoFilterBackend, OrderingFilter)
+    filter_class = CourseOverviewFilter
+    search_fields = ['display_name', 'id']
+    ordering_fields = ['display_name', 'self_paced']
```
Comment (OmarIthawi) [inline figures/views.py]:
```suggestion
    ordering_fields = ['display_name', 'self_paced', 'date_joined']
```

File: figures/views.py [hunk 4]
```diff
@@ -357,8 +365,10 @@ class GeneralUserDataViewSet(CommonAuthMixin, viewsets.ReadOnlyModelViewSet):
     model = get_user_model()
     pagination_class = FiguresKiloPagination
     serializer_class = GeneralUserDataSerializer
-    filter_backends = (DjangoFilterBackend, )
+    filter_backends = (SearchFilter, DjangoFilterBackend, OrderingFilter)
     filter_class = UserFilterSet
+    search_fields = ['username', 'email', 'profile__name']
```
Comment (OmarIthawi) [inline figures/views.py]:
```suggestion
    search_fields = ['username', 'email', 'name']
```

File: figures/views.py [hunk 5]
```diff
@@ -357,8 +365,10 @@ class GeneralUserDataViewSet(CommonAuthMixin, viewsets.ReadOnlyModelViewSet):
...
+    ordering_fields = ['username', 'email', 'profile__name', 'is_active']
```
Comment (OmarIthawi) [inline figures/views.py]:
```suggestion
    ordering_fields = ['username', 'email', 'name', 'is_active', 'date_joined']
```

File: figures/filters.py [hunk 6]
```diff
@@ -124,7 +125,7 @@ class UserFilterSet(django_filters.FilterSet):
...
                   'is_superuser', 'enrolled_in_course_id', 'user_ids', ]
```
Comment (OmarIthawi) [inline figures/filters.py]:
```suggestion
                  'is_superuser', 'enrolled_in_course_id', 'user_ids', 'date_joined']
```

File: tests/views/test_general_course_data_view.py [hunk 7]
```diff
@@ -226,3 +236,21 @@ def test_get_with_course_id_for_other_site(self):
             view = self.view_class.as_view({'get': 'retrieve'})
             response = view(request, pk=str(course.id))
             assert response.status_code == 403
+
+    def test_get_search(self):
```
Comment (OmarIthawi) [inline tests/views/test_general_course_data_view.py]: There's a couple of [good reasons for not to write `for`-loops inside tests](https://automationrhapsody.com/data-driven-testing-junit-parameterized-tests/). I won't expand here, but give the alternative:

```suggestion
    @pytest.mark.parametrize('search_term', SEARCH_TERMS)
    def test_get_search(self, search_term):
```

File: tests/views/test_general_course_data_view.py [hunk 8]
```diff
@@ -226,3 +236,21 @@ def test_get_with_course_id_for_other_site(self):
...
+        """
+        Based on a SEARCH_TERMS data set, we query the endpoint with search
+        terms and we compare with the expected results.
+        """
+        for search_term in SEARCH_TERMS:
```
Comment (OmarIthawi) [inline tests/views/test_general_course_data_view.py]: Same as above:

```suggestion
        request_path = self.request_path + '?search=' + search_term['term']
        request = APIRequestFactory().get(request_path)
        force_authenticate(request, user=self.staff_user)
        view = self.view_class.as_view({'get': 'list'})
        response = view(request)

        assert response.status_code == 200
        assert response.data['count'] == search_term['expected_result']
        assert len(response.data['results']) == search_term['expected_result']
```

File: tests/views/test_general_user_data_view.py [hunk 9]
```diff
@@ -185,3 +192,21 @@ def get_course_rec(course_id, course_list):
             for course_enrollment in CourseEnrollment.objects.filter(user=user_model):
                 # Test that the course id exists in the data
                 assert get_course_rec(course_enrollment.course_id, rec['courses'])
+
+    def test_get_search(self):
```
Comment (OmarIthawi) [inline tests/views/test_general_user_data_view.py]:
```suggestion
   @pytest.mark.parametrize('search_term', SEARCH_TERMS)
    def test_get_search(self, search_term):
```

File: tests/views/test_general_user_data_view.py [hunk 10]
```diff
@@ -185,3 +192,21 @@ def get_course_rec(course_id, course_list):
...
+        """
+        Based on a SEARCH_TERMS data set, we query the endpoint with search
+        terms and we compare with the expected results.
+        """
+        for search_term in SEARCH_TERMS:
+            request_path = self.request_path + '?search=' + search_term['term']
+            print request_path
```
Comment (OmarIthawi) [inline tests/views/test_general_user_data_view.py]:
```suggestion
```

**Comment by johnbaldwin** [comment]
@melvinsoft Looks like tests are failing on not recognizing `self_paced` I can look into this in the morning. You are working with some of my earliest test code I believe, as I was figuring out what I was doing, so it is not the easiest test code to work with. So glad to see you working on Figures!!!

**Comment by johnbaldwin** [comment]
@melvinsoft The tests are failing because the CourseOverviewFactory construction helper function is now trying to get `self_paced` as a dict key, but `self_paced` does not exist in the kwargs: https://github.com/appsembler/figures/blob/maxi/add-search-and-filtering/tests/views/test_general_course_data_view.py#L85

The source data are here: https://github.com/appsembler/figures/blob/maxi/add-search-and-filtering/tests/views/test_general_course_data_view.py#L52-L57

One example where it is called is here: https://github.com/appsembler/figures/blob/maxi/add-search-and-filtering/tests/views/test_general_course_data_view.py#L112


If you do not need to specify `self_paced` as `True` or `False` for specific tests, you can simply remove it from the arguments here: https://github.com/appsembler/figures/blob/maxi/add-search-and-filtering/tests/views/test_general_course_data_view.py#L85 and accept the default value.

If you do need to test with different values, then one option is to add `self_paced` to the values here: https://github.com/appsembler/figures/blob/maxi/add-search-and-filtering/tests/views/test_general_course_data_view.py#L52-L57

File: figures/filters.py (hunk 1 above)
Comment (melvinsoft) [reply figures/filters.py]: Done!

File: figures/views.py (hunk 4 above)
Comment (melvinsoft) [reply figures/views.py]: I cannot change this,  the search fields names need to be for valid model fields, it cannot be serializer fields, so I need to use the look up.

File: figures/views.py (hunk 5 above)
Comment (melvinsoft) [reply figures/views.py]: Done.

**Comment by melvinsoft** [comment]
@OmarIthawi @johnbaldwin Thanks you both for your reviews! I've the tests fixed now, and all the checks are passing!  Please give it another look.

File: tests/views/test_general_user_data_view.py [hunk 11]
```diff
@@ -193,14 +193,14 @@ def get_course_rec(course_id, course_list):
...
                 assert get_course_rec(course_enrollment.course_id, rec['courses'])
 
-    def test_get_search(self):
+    @pytest.mark.parametrize('search_term', SEARCH_TERMS)
+    def test_get_search(self, search_term):
         """
         Based on a SEARCH_TERMS data set, we query the endpoint with search
         terms and we compare with the expected results.
         """
         for search_term in SEARCH_TERMS:
```
Comment (OmarIthawi) [inline tests/views/test_general_user_data_view.py]:
```suggestion
```

Comment (melvinsoft) [reply tests/views/test_general_user_data_view.py]: Thanks @OmarIthawi ! I totally missed this one!

**Comment by melvinsoft** [comment]
@OmarIthawi Thanks again for the last catch, I think it's ready now.

File: tests/views/test_general_user_data_view.py [hunk 12]
```diff
@@ -185,3 +193,27 @@ def get_course_rec(course_id, course_list):
...
+        Based on a SEARCH_TERMS data set, we query the endpoint with search
+        terms and we compare with the expected results.
+        """
+        request_path = self.request_path + '?search=' + search_term['term']
+        request = APIRequestFactory().get(request_path)
+        force_authenticate(request, user=self.staff_user)
+        view = self.view_class.as_view({'get': 'list'})
+        response = view(request)
+        assert response.status_code == 200
+        if not is_multisite():
```
Comment (OmarIthawi) [inline tests/views/test_general_user_data_view.py]: This is a cosmetic change that can be addressed on a future PR. But generally an `if` statement, means that there are two separate test cases. 

```python
    @pytest.mark.skipif(organizations_support_sites(),
                        reason='Organizations support sites')
    @pytest.mark.parametrize('search_term', SEARCH_TERMS)
    def test_get_search(self, search_term):
        """
        Based on a SEARCH_TERMS data set, we query the endpoint with search
        terms and we compare with the expected results.
        """
        request_path = self.request_path + '?search=' + search_term['term']
        request = APIRequestFactory().get(request_path)
        force_authenticate(request, user=self.staff_user)
        view = self.view_class.as_view({'get': 'list'})
        response = view(request)
        assert response.status_code == 200
        assert response.data['count'] == search_term['expected_result']
        assert len(response.data['results']) == search_term['expected_result']

    @pytest.mark.skipif(not organizations_support_sites(),
                        reason='Organizations does not support sites')
    @pytest.mark.parametrize('search_term', SEARCH_TERMS)
    def test_get_search(self, search_term):
        """
        Based on a SEARCH_TERMS data set, we query the endpoint with search
        terms and we compare with the expected results.
        """
        request_path = self.request_path + '?search=' + search_term['term']
        request = APIRequestFactory().get(request_path)
        force_authenticate(request, user=self.staff_user)
        view = self.view_class.as_view({'get': 'list'})
        response = view(request)
        assert response.status_code == 200
        # the defaul site is example.com so it won't be find users, and
        # that the expected outcome, we should test with different
        # sites expecting different results, but is out of scope for
        # now.
        assert response.data['count'] == 0
        assert len(response.data['results']) == 0
```

Comment (melvinsoft) [reply tests/views/test_general_user_data_view.py]: 👍 Ok, this is good to know, but please, let's address it in a future PR.

**Comment by grozdanowski** [comment]
@johnbaldwin can you take a last look and merge into master as quick as possible? I want to wrap up the frontend (hook it up to this) tomorrow ideally.

File: tests/views/test_general_user_data_view.py (hunk 12 above)
Comment (OmarIthawi) [reply tests/views/test_general_user_data_view.py]: no problem maxi!

Comment (johnbaldwin) [reply tests/views/test_general_user_data_view.py]: This is something that we need to figures out, how we want to run tests that are single site versus multisite. I've done both checking for multisite in order to wire up `UserOrganizationMapping` and constructing `OrganizationCourse` with and without the `sites=[some_site]` parameter. Something that would be a good investigative and decision topic for the unification effort

**Pull request merged**